CylindricalEndcap: Allow target sphere to be inside source sphere.

markscheel opened the pull request:
## Proposed changes

CylindricalEndcap is a map that maps a 3d unit right cylinder to the portion of a volume contained between two spheres, where the volume is bounded by the intersection of the 'source' sphere with a z=constant plane.
For convenience and simplicity, CylindricalEndcap demands that the 'target' sphere contains the 'source' sphere.
However, it turns out that for the cylindrical BBH domain we also need CylindricalEndcap for the case in which the 'source' sphere contains the 'target' sphere.

This PR modifies CylindricalEndcap so that it allows the case in which the 'source' sphere contains the 'target' sphere.
The changes necessary were:
  - Change the logic choosing a particular root of a quadratic equation in the formula for the map and its inverse; the correct root to choose depends on the relationship between the two spheres.
  - Change the sanity checks in the CylindricalEndcap constructor, which now allow either sphere to be contained in the other.
  - Add the newly allowed configurations to Test_CylindricalEndcap.cpp.
  - Simplify the existing test in Test_CylindricalEndcap.cpp and make it (and the matching sanity checks in CylindricalEndcap.cpp) more robust.
  - Update the documentation.
  
 This should fix #2879

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
FocallyLifted: Add source_is_between_focus_and_target to lamtil.

markscheel committed:
CylindricalEndcap: Allow sphere2 contained in sphere1.

Previously we demanded that sphere2 contains sphere1.
Now we demand either that sphere2 contains sphere1 or sphere1
contains sphere2.

Also, some of the sanity checks in the constructor still allowed
rare cases where the unit tests fail by more than
100 times machine roundoff.  Those sanity checks have now been
adjusted so that the tests are more robust.

markscheel committed:
FocallyLiftedMap: Add source_is_between_focus_and_target.

Previously this parameter was a static member function of
the InnerMap.  But it turns out that the same InnerMap can
be used both for cases when the source is between the focus and
the target, and when it is not.  So now this parameter is a bool
passed into the constructor.

markscheel commented:
Rebased onto develop and fixed many conflicts.

markscheel commented:
Updated the last commit to fix a gcc compiler warning.

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedMap.hpp:
```diff
@@ -99,6 +99,17 @@ bool operator==(const FocallyLiftedMap<InnerMap>& lhs,
+ * root, corresponding to two intersections of the sphere.  The
+ * boolean parameter `source_is_between_focus_and_target` that is
+ * passed into the constructor of `FocallyLiftedMap` is used to choose the
+ * appropriate root, or to die with an error messages if a suitable
```
maybe `or to error if a suitable...` would be less dark :)

nilsdeppe commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp:
```diff
@@ -320,15 +320,20 @@ void Endcap::dxbar_dsigma(
+  // If the target surface is outside the sphere (that is,
+  // source_is_between_focus_and_target is true), then lambda_tilde should be
+  // positive and less than or equal to unity. If the target surface is inside
+  // the sphere, then lambda_tilde should be positive and greater than or equal
```
would `lambda_tilde must be greater than equal equal to one` by clearer?

nilsdeppe reviewed (dismissed):
A couple minor suggestions, please squash if you think they're worth doing

markscheel committed:
Test_CylindricalEndcap:  Test sphere2 contained in sphere1

 - Add a new test that tests the case when sphere2 is contained
   inside sphere1.

 - Simplifies the existing test (the case when sphere1 is contained
   in sphere2)

 - Make the existing test more robust by adjusting some of the
   simplifying assumptions.

markscheel commented:
Rebased on develop to resolve conflict.

moxcodes commented on src/Domain/CoordinateMaps/FocallyLiftedEndcap.hpp:
```diff
@@ -436,10 +436,6 @@ class Endcap {
 
```
commit message: missing a space in `bothfor`

markscheel replied:
fixed

moxcodes commented on src/Domain/CoordinateMaps/CylindricalEndcap.cpp:
```diff
@@ -110,65 +126,103 @@ CylindricalEndcap::CylindricalEndcap(const std::array<double, 3>& center_one,
-  ASSERT(tan_beta < tan(max_opening_angle),
-         "Opening angle is too large. The map has not "
-         "been tested for this case.");
+    // Check if opening angle is small enough.
```
It appears that the opening angle condition is only enforced for the case where sphere1 is contained in sphere2. Is that intentional? Why is the condition relaxed for the alternative case?

markscheel replied:
Yes, it's intentional. I added a comment.

moxcodes commented on src/Domain/CoordinateMaps/CylindricalEndcap.hpp:
```diff
@@ -62,7 +62,8 @@ namespace domain::CoordinateMaps {
  *
- * We demand that Sphere 1 is fully contained inside Sphere 2. It is
+ * We demand that either Sphere 1 is fully contained inside Sphere 2, or
+ * that Sphere 2 is fully contained inside Sphere 1. It is
```
On a first approach to this map, it's not necessarily clear what the ramifications are of putting sphere2 inside sphere1 instead. Is it effectively the same map, but the lines drawn from the projection point to the spheres now meet at the points where the plane intersects the outer sphere instead? Is there any reversal of coordinates that the user should be aware of when interchanging sphere 1 and sphere 2 like that?

markscheel replied:
Only Sphere1 has its mapped region bounded by a plane.  To construct Domains out of multiple Blocks, each with their own CylindricalEndcaps, you need touching Blocks to share the same boundaries, which in some cases are the intersection of a sphere and a plane.

I added more text to the dox to try to explain this, using the cylindrical BBH in the 2012 paper as an example.

markscheel replied:
And in all cases you need to worry about reversing/rotating coordinates. CylindricalEndcap is by default oriented in the z direction but in the BBH Domain we will want corresponding Blocks oriented in +x and -x, so in the DomainCreator there will be different OrientationMaps applied to different Blocks.

moxcodes reviewed (commented):
Looks good, only a couple small comments and suggestions.
I didn't go through and verify all of the numerical choices for bounds on the endcap parameters -- I'll trust that if those aren't working acceptably, the test would at least occasionally fail.

markscheel commented:
Rebased, fixed typo in commit message, and added a few fixups addressing questions of @moxcodes

moxcodes commented:
looks good, thanks for clarifying those points. I'm happy for you to go ahead and squash.

markscheel commented:
squashed.
